File: 04-Data-Structures/code/typescript/src/binaryHeap.ts
import { DSError } from "./errors.js";

/**
 * Binary min-heap over an array. `toList()` returns the raw heap-ordered
 * backing array (a valid heap, not a sorted sequence) -- vectors should
 * prefer `peek`/`pop` sequences to assert ordering.
 */
export class BinaryHeap<T = number> {
  private data: T[] = [];
  private cmp: (a: T, b: T) => number;

  constructor(cmp: (a: T, b: T) => number = (a, b) => (a as unknown as number) - (b as unknown as number)) {
    this.cmp = cmp;
  }

  size(): number {
    return this.data.length;
  }

  peek(): T {
    if (this.data.length === 0) throw new DSError("empty", "peek from empty heap");
    return this.data[0];
  }

  push(x: T): void {
    this.data.push(x);
    this.bubbleUp(this.data.length - 1);
  }

  pop(): T {
    if (this.data.length === 0) throw new DSError("empty", "pop from empty heap");
    const top = this.data[0];
    const last = this.data.pop() as T;
    if (this.data.length > 0) {
      this.data[0] = last;
      this.bubbleDown(0);
    }
    return top;
  }

  toList(): T[] {
    return [...this.data];
  }

  private bubbleUp(i: number): void {
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.cmp(this.data[i], this.data[parent]) < 0) {
        [this.data[i], this.data[parent]] = [this.data[parent], this.data[i]];
        i = parent;
      } else break;
    }
  }

  private bubbleDown(i: number): void {
    const n = this.data.length;
    while (true) {
      let smallest = i;
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      if (l < n && this.cmp(this.data[l], this.data[smallest]) < 0) smallest = l;
      if (r < n && this.cmp(this.data[r], this.data[smallest]) < 0) smallest = r;
      if (smallest === i) break;
      [this.data[i], this.data[smallest]] = [this.data[smallest], this.data[i]];
      i = smallest;
    }
  }

  checkInvariants(): void {
    for (let i = 1; i < this.data.length; i++) {
      const parent = (i - 1) >> 1;
      if (this.cmp(this.data[i], this.data[parent]) < 0) throw new DSError("invalid", "heap property violated");
    }
  }
}

File: 04-Data-Structures/code/python/seb_ds/binary_heap.py
"""Binary min-heap over an array. Mirrors `typescript/src/binaryHeap.ts`.

`to_list()` returns the raw heap-ordered backing list (a valid heap, not a
sorted sequence) -- vectors should prefer `peek`/`pop` sequences to assert
ordering.
"""

from __future__ import annotations

from typing import Callable, Generic, List, Optional, TypeVar

from .errors import DSError

T = TypeVar("T")


def _default_cmp(a: T, b: T) -> int:
    return (a > b) - (a < b)  # type: ignore[operator]


class BinaryHeap(Generic[T]):
    def __init__(self, cmp: Optional[Callable[[T, T], int]] = None) -> None:
        self._data: List[T] = []
        self._cmp = cmp or _default_cmp

    def size(self) -> int:
        return len(self._data)

    def peek(self) -> T:
        if not self._data:
            raise DSError("empty", "peek from empty heap")
        return self._data[0]

    def push(self, x: T) -> None:
        self._data.append(x)
        self._bubble_up(len(self._data) - 1)

    def pop(self) -> T:
        if not self._data:
            raise DSError("empty", "pop from empty heap")
        top = self._data[0]
        last = self._data.pop()
        if self._data:
            self._data[0] = last
            self._bubble_down(0)
        return top

    def to_list(self) -> List[T]:
        return list(self._data)

    def _bubble_up(self, i: int) -> None:
        while i > 0:
            parent = (i - 1) >> 1
            if self._cmp(self._data[i], self._data[parent]) < 0:
                self._data[i], self._data[parent] = self._data[parent], self._data[i]
                i = parent
            else:
                break

    def _bubble_down(self, i: int) -> None:
        n = len(self._data)
        while True:
            smallest = i
            l = 2 * i + 1
            r = 2 * i + 2
            if l < n and self._cmp(self._data[l], self._data[smallest]) < 0:
                smallest = l
            if r < n and self._cmp(self._data[r], self._data[smallest]) < 0:
                smallest = r
            if smallest == i:
                break
            self._data[i], self._data[smallest] = self._data[smallest], self._data[i]
            i = smallest

    def check_invariants(self) -> None:
        for i in range(1, len(self._data)):
            parent = (i - 1) >> 1
            if self._cmp(self._data[i], self._data[parent]) < 0:
                raise DSError("invalid", "heap property violated")
